### PR TITLE
GH-37014: [C++][Parquet] Preserve some Parquet distinct counts when merging stats

### LINKED ISSUE #37014: [C++] Merging Parquet stats can preserve distinct counts when either side is zero
### Describe the enhancement requested

https://github.com/apache/arrow/blob/710f96069f9f036338a0dd99202f0486bb7d1984/cpp/src/parquet/statistics.cc#L565 is more conservative than necessary. While in general we can't preserve distinct counts, since we don't know the set of overlapping values, we can preserve distinct counts if either `this` or `other` has a distinct count of 0.

I'm preparing a PR, assuming this sounds reasonable to fix.

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by sfc-gh-ebrossard**
### Rationale for this change

Statistics merging is unnecessarily conservative with respect to preserving distinct counts. If either side is 0, the merge stats can preserve the distinct count, instead of clearing it.

### What changes are included in this PR?

Distinct count is preserved if either the lhs or rhs has a distinct count of 0 when merging stats.

### Are these changes tested?

Yes, added a unit test.

### Are there any user-facing changes?

I don't think this change to stats is considered to be user-facing.
* Closes: #37014

### CODE REVIEW

File: cpp/src/parquet/statistics_test.cc [hunk 1]
```diff
@@ -645,6 +645,50 @@ class TestStatisticsHasFlag : public TestStatistics<TestType> {
...
+      statistics4 = std::dynamic_pointer_cast<TypedStatistics<TestType>>(
+          Statistics::Make(this->schema_.Column(0), &encoded_statistics4,
+                           /*num_values=*/10));
+      EXPECT_TRUE(statistics4->HasDistinctCount());
+    }
+
+    // Both sides have 0 for the distinct count.
+    VerifyMergedStatistics(*statistics3, *statistics3,
+                           [](TypedStatistics<TestType>* merged_statistics) {
+                             EXPECT_TRUE(merged_statistics->HasDistinctCount());
```
Comment (sfc-gh-ebrossard) [inline cpp/src/parquet/statistics_test.cc:674]: Sorry, these tests are still failing. I'll see if I can get them running locally with https://arrow.apache.org/docs/developers/cpp/development.html#running-unit-tests and figure out what I've missed here.

File: cpp/src/parquet/statistics.cc [hunk 2]
```diff
@@ -621,7 +628,9 @@ class TypedStatisticsImpl : public TypedStatistics<DType> {
       // num_values_ is reliable and it means number of non-null values.
       s.all_null_value = num_values_ == 0;
     }
-    // TODO (GH-36505): distinct count is not encoded for now.
+    if (HasDistinctCount()) {
```
Comment (sfc-gh-ebrossard) [inline cpp/src/parquet/statistics.cc:631]: @mapleFU FYI I noticed that distinct counts aren't propagated to `EncodedStatistics`, as tracked in https://github.com/apache/arrow/issues/36505. For my new test, I could potentially just check `TypedStatistics` without calling `Encode()`, if you'd rather that I leave this as-is.

**Comment by mapleFU** [comment]
I'm OK with the change, but seems that `distinct_count == 0` only happens when the page is all null? Is it a common case?

And encode `distinct_count` seems not applied by parquet-mr and rust impl.

**Comment by sfc-gh-ebrossard** [comment]
> I'm OK with the change, but seems that `distinct_count == 0` only happens when the page is all null? Is it a common case?
> 
> And encode `distinct_count` seems not applied by parquet-mr and rust impl.

Yeah, the case I was seeing is that we merged stats from an all-null page with a page that had non-null values and ended up with an unset distinct count. I thought it would be useful to preserve the count if we can.

Technically we could make this more general and propagate distinct counts for pages whose min and max ranges don't overlap, too. For example, if one page has a range of `[0, 9]` and the other has a range of `[10, 19]`, we can simply add the distinct counts. What do you think? Maybe that would add too much complexity, though.

Do you mind pointing me to the code for parquet-ml and rust that I should update? If this change looks generally okay for the C++ Parquet code, I'll work on those as well.

**Comment by mapleFU** [comment]
Personally, I thin `distinct_count` might only usable when using **dictionary encoding**. And using it would be bug-prune.

Before my patch https://github.com/apache/arrow/pull/35989/files , the `distinct_count` is directly added with another. So I think the distinct_count is always misused.

I think your commit is reasonable, but also I think using `distinct_count` is a bit dangerous. Maybe add a `SetDistinctCount` and maintaining it outside would be better? Personally, I use `SetDistinctCount` when building the statistics of a dictionary-encoded non-fallbacked ColumnChunk

Also cc @pitrou @wgtmac

File: cpp/src/parquet/statistics.cc (hunk 2 above)
Comment (mapleFU) [reply cpp/src/parquet/statistics.cc:631]: Yes, previously I think we can write it, but pirtou mentioned that it's bug-prone, and other implemention didn't write that. So I set it not encoded.

**Comment by mapleFU** [comment]
1. https://github.com/apache/arrow-rs/blob/841a6a981dfa7ece0369b1fc8be77555f036c1b4/parquet/src/file/statistics.rs#L231 Rust write distinct count. But it will not merge that. The merge logic is consistent with our
2. parquet-mr seems doesn't write it

File: cpp/src/parquet/statistics_test.cc [hunk 3]
```diff
@@ -602,18 +602,25 @@ class TestStatisticsHasFlag : public TestStatistics<TestType> {
...
     return chunk_statistics;
   }
 
   void VerifyMergedStatistics(
       const TypedStatistics<TestType>& stats1, const TypedStatistics<TestType>& stats2,
-      const std::function<void(TypedStatistics<TestType>*)>& test_fn) {
-    ASSERT_NO_FATAL_FAILURE(test_fn(MergedStatistics(stats1, stats2).get()));
-    ASSERT_NO_FATAL_FAILURE(test_fn(MergedStatistics(stats2, stats1).get()));
+      const std::function<void(TypedStatistics<TestType>*)>& test_fn,
+      const EncodedStatistics* initial_statistics = nullptr) {
```
Comment (pitrou) [inline cpp/src/parquet/statistics_test.cc:619]: The signature is a bit weird here. How about two different overloads?
```c++

  void VerifyMergedStatistics(
      const TypedStatistics<TestType>& stats1, const TypedStatistics<TestType>& stats2,
      const std::function<void(TypedStatistics<TestType>*)>& test_fn) { ... }
  void VerifyMergedStatistics(
      const EncodedStatistics& initial_statistics,
      const TypedStatistics<TestType>& stats1, const TypedStatistics<TestType>& stats2,
      const std::function<void(TypedStatistics<TestType>*)>& test_fn) { ... }
```

File: cpp/src/parquet/statistics_test.cc [hunk 4]
```diff
@@ -645,6 +652,50 @@ class TestStatisticsHasFlag : public TestStatistics<TestType> {
...
+      EXPECT_TRUE(statistics3->HasDistinctCount());
+
+      EncodedStatistics encoded_statistics4;
+      encoded_statistics4.has_distinct_count = true;
+      encoded_statistics4.distinct_count = 10;
+      statistics4 = std::dynamic_pointer_cast<TypedStatistics<TestType>>(
+          Statistics::Make(this->schema_.Column(0), &encoded_statistics4,
+                           /*num_values=*/10));
+      EXPECT_TRUE(statistics4->HasDistinctCount());
+    }
```
Comment (pitrou) [inline cpp/src/parquet/statistics_test.cc:676]: There starts to be a lot of copy-pasted code here. How about defining specific helpers? For example:
```c++
std::optional<int64_t> MergeDistinctCount(std::optional<int64_t> initial, std::vector<std::optional<int64_t>> subsequent)
```

And then
```c++
ASSERT_EQ(std::nullopt, MergeDistinctCount(std::nullopt, {0});
ASSERT_EQ(std::nullopt, MergeDistinctCount(std::nullopt, {10, 0});
ASSERT_EQ({10}, MergeDistinctCount({10}, {0, 0});
ASSERT_EQ({10}, MergeDistinctCount({0}, {10, 0});
ASSERT_EQ({10}, MergeDistinctCount({0}, {0, 10});
ASSERT_EQ(std::nullopt, MergeDistinctCount({10}, {0, 10});
ASSERT_EQ(std::nullopt, MergeDistinctCount({10}, {0, std::nullopt});
// etc?
```

**Comment by mapleFU** [comment]
@sfc-gh-ebrossard Hi, overall this patch looks ok for me, would you like to move forward

File: cpp/src/parquet/statistics_test.cc (hunk 4 above)
Comment (sfc-gh-ebrossard) [reply cpp/src/parquet/statistics_test.cc]: Added a `MergeDistinctCount` test helper; PTAL.

File: cpp/src/parquet/statistics_test.cc (hunk 3 above)
Comment (sfc-gh-ebrossard) [reply cpp/src/parquet/statistics_test.cc]: Now we just use `MergeDistinctCount` for the distinct count tests, so I removed the overload.

File: cpp/src/parquet/statistics.cc (hunk 2 above)
Comment (sfc-gh-ebrossard) [reply cpp/src/parquet/statistics.cc:631]: Sorry, by, "That seems ok to me," do you mean to leave this here? Or to remove it?

**Comment by sfc-gh-ebrossard** [comment]
> @sfc-gh-ebrossard Hi, overall this patch looks ok for me, would you like to move forward

Thanks for the nudge; I addressed the comments :)

**Review by pitrou — APPROVED** [review]
+1, thanks for the update @sfc-gh-ebrossard !

**Pull request merged**